## Part 1: Preprocessing

In [19]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [20]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [21]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[["Attrition", "Department"]]
y_df

Attrition              Department
0          Yes                   Sales
1           No  Research & Development
2          Yes  Research & Development
3           No  Research & Development
4           No  Research & Development
...        ...                     ...
1465        No  Research & Development
1466        No  Research & Development
1467        No  Research & Development
1468        No                   Sales
1469        No  Research & Development

[1470 rows x 2 columns]

In [22]:
# Create a list of at least 10 column names to use as X data
list_x = ["Education",
         "Age",
           "DistanceFromHome",
             "JobSatisfaction",
               "HourlyRate",
                 "StockOptionLevel",
                   "WorkLifeBalance",
                     "YearsAtCompany",
                       "YearsSinceLastPromotion",
                         "NumCompaniesWorked"]


# Create X_df using your selected columns
X_df = attrition_df[list_x]

# Show the data types for X_df
X_df.dtypes

Education                  int64
Age                        int64
DistanceFromHome           int64
JobSatisfaction            int64
HourlyRate                 int64
StockOptionLevel           int64
WorkLifeBalance            int64
YearsAtCompany             int64
YearsSinceLastPromotion    int64
NumCompaniesWorked         int64
dtype: object

In [23]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=42)

In [ ]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary

# NOTE This step is no longer necessary as I opted to change "OverTime" (a non-numeric column)
# to HourlyRate (a numeric column). Now everything is already numeric. I left my code
# (which wasn't working the way I wanted) below for future review

# Define a mapping dictionary
#mapping_dict = {'Yes': 1, 'No': 0}

# Convert 'OverTime' column to numeric
#X_train['OverTime'] = X_train['OverTime'].map(mapping_dict)
#X_test['OverTime'] = X_test['OverTime'].map(mapping_dict)

In [24]:
# Check the data types
print(X_train.dtypes)

# Display the first few rows
print(X_train.head())

Education                  int64
Age                        int64
DistanceFromHome           int64
JobSatisfaction            int64
HourlyRate                 int64
StockOptionLevel           int64
WorkLifeBalance            int64
YearsAtCompany             int64
YearsSinceLastPromotion    int64
NumCompaniesWorked         int64
dtype: object
      Education  Age  DistanceFromHome  JobSatisfaction  HourlyRate  \
1343          3   29                 7                1          59   
1121          4   36                 1                3          73   
1048          3   34                 3                1          81   
1393          3   27                 9                4          44   
527           3   32                10                4          55   

      StockOptionLevel  WorkLifeBalance  YearsAtCompany  \
1343                 0                3               3   
1121                 0                3               1   
1048                 0                3             

In [25]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_scaler = scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [26]:
from sklearn.preprocessing import OneHotEncoder
# Create a OneHotEncoder for the Department column
department_encoder = OneHotEncoder(sparse=False)

# Fit the encoder to the training data
department_encoder.fit(y_train[['Department']])

# Create two new variables by applying the encoder to the training and testing data
y_train_department = department_encoder.transform(y_train[['Department']])
y_test_department = department_encoder.transform(y_test[['Department']])

c:\Users\brand\anaconda3\Lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [27]:
# Create a OneHotEncoder for the Attrition column
attrition_encoder = OneHotEncoder(sparse=False)

# Fit the encoder to the training data
attrition_encoder.fit(y_train[['Attrition']])

# Create two new variables by applying the encoder to the training and testing data
y_train_attrition = attrition_encoder.transform(y_train[['Attrition']])
y_test_attrition = attrition_encoder.transform(y_test[['Attrition']])

c:\Users\brand\anaconda3\Lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


## Create, Compile, and Train the Model

In [28]:
# Find the number of columns in the X training data
num_columns = X_train.shape[1]
print("Number of columns in the X training data: ", num_columns)

# Create the input layer
input_layer = layers.Input(shape=(10,), name='input_features')

# Create at least two shared layers
shared_layer1 = layers.Dense(64, activation='relu')(input_layer)
shared_layer2 = layers.Dense(128, activation='relu')(shared_layer1)

Number of columns in the X training data:  10


In [29]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
department_hidden = layers.Dense(32, activation='relu')(shared_layer2)

# Create the output layer
# The number of nodes should match the number of unique values in the 'Department' column
# so we can do a quick check for that
num_unique_departments = len(y_train['Department'].unique())
department_output = layers.Dense(num_unique_departments, activation='softmax', name='department_output')(department_hidden)

In [30]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden = layers.Dense(32, activation='relu')(shared_layer2)

# Create the output layer
# The number of nodes should match the number of unique values in the 'Attrition' column
# so we can do a quick check for that
num_unique_attritions = len(y_train['Attrition'].unique())
attrition_output = layers.Dense(num_unique_attritions, activation='sigmoid', name='attrition_output')(attrition_hidden)

In [31]:
# Create the model
model = Model(inputs=input_layer, outputs=[department_output, attrition_output])

# Compile the model
model.compile(optimizer='adam',
              loss={'department_output': 'categorical_crossentropy', 'attrition_output': 'categorical_crossentropy'},
              metrics={'department_output': 'accuracy', 'attrition_output': 'accuracy'})

# Summarize the model
model.summary()

Model: "model_1"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_features (InputLayer  [(None, 10)]                 0         []                            
 )                                                                                                
                                                                                                  
 dense_4 (Dense)             (None, 64)                   704       ['input_features[0][0]']      
                                                                                                  
 dense_5 (Dense)             (None, 128)                  8320      ['dense_4[0][0]']             
                                                                                                  
 dense_6 (Dense)             (None, 32)                   4128      ['dense_5[0][0]']       

In [32]:
# Train the model
model.fit(
    X_train_scaled,
    {'department_output': y_train_department, 'attrition_output': y_train_attrition},
    epochs=10,
    batch_size=32,
    validation_split=0.2
)


Epoch 1/10
28/28 [==============================] - 1s 9ms/step - loss: 1.3589 - department_output_loss: 0.8303 - attrition_output_loss: 0.5286 - department_output_accuracy: 0.6572 - attrition_output_accuracy: 0.7866 - val_loss: 1.4309 - val_department_output_loss: 0.8605 - val_attrition_output_loss: 0.5704 - val_department_output_accuracy: 0.6063 - val_attrition_output_accuracy: 0.7873
Epoch 2/10
28/28 [==============================] - 0s 2ms/step - loss: 1.2039 - department_output_loss: 0.7631 - attrition_output_loss: 0.4408 - department_output_accuracy: 0.6674 - attrition_output_accuracy: 0.8388 - val_loss: 1.3635 - val_department_output_loss: 0.8492 - val_attrition_output_loss: 0.5143 - val_department_output_accuracy: 0.6063 - val_attrition_output_accuracy: 0.7873
Epoch 3/10
28/28 [==============================] - 0s 2ms/step - loss: 1.1652 - department_output_loss: 0.7514 - attrition_output_loss: 0.4138 - department_output_accuracy: 0.6663 - attrition_output_accuracy: 0.8388 - v

In [33]:
# Evaluate the model with the testing data
results = model.evaluate(X_test_scaled, {'department_output': y_test_department, 'attrition_output': y_test_attrition})

print("Test loss, Test accuracy:", results)

12/12 [==============================] - 0s 1ms/step - loss: 1.1910 - department_output_loss: 0.8149 - attrition_output_loss: 0.3761 - department_output_accuracy: 0.6495 - attrition_output_accuracy: 0.8750
Test loss, Test accuracy: [1.19098699092865, 0.8148515820503235, 0.3761354386806488, 0.6494565010070801, 0.875]


In [34]:
# Print the accuracy for both department and attrition
print(f"Department Accuracy: {results[3]}")
print(f"Attrition Accuracy: {results[4]}")

Department Accuracy: 0.6494565010070801
Attrition Accuracy: 0.875


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. Accuracy is a commonly used metric in classification algorithms such as this, but may not be suitable if the data is asymmetric. For instance, if there are a very few number of employees who leave the company in the data, then the accuracy of the Attrition metric may be meaningless. Likewise, if the distribution of employees across different departments isn't uniform, then the model may be skewed towards whichever department has the most data. Refer to the code cell located below this markdown, where I decided to further explore this question. As you can see, Attrition is skewed heavily towards "no", and Department is skewed heavily towards "Research & Development". This could make the model biased towards these values. You would need to implement weighted accuracy as a metric for the model to take into account.

2. For the activation functions of the output layers, softmax was used for Department while sigmoid was used for Attrition. This is because Department was a categorical variable with multiple categories, and the multi-class softmax was the most suitable for this. Meanwhile, Attrition was a simple binary "yes" or "no", and sigmoid is the activation function of choice for binary classification problems.

3. This model can be improved in a few ways: adding more layers, including more variables, or collecting more data.

    Layers: We only have a few layers in this neural network. We can increase the complexity by adding more layers and more epochs. While this would increase training time, it would potentially lead to a more accurate model. We could also spend time tuning hyperparameters, or trying different optimizers. 

    Variables: The X data only used the 10 columns I selected. This model could potentially be improved by adding more or all of the columns.

    Data: There are only 1470 employees in this dataset. While that seems like a lot, collecting more could give the model more to learn from and lead to a more accurate model.

In [35]:
# 1. Is accuracy the best metric to use on this data? Why or why not?

# Summary of the 'Attrition' column
attrition_summary = attrition_df['Attrition'].value_counts()
print("Attrition summary:\n", attrition_summary)

# Summary of the 'Department' column
department_summary = attrition_df['Department'].value_counts()
print("\nDepartment summary:\n", department_summary)

Attrition summary:
 Attrition
No     1233
Yes     237
Name: count, dtype: int64

Department summary:
 Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64
